framework/doc/content/source/constraints/PenaltyPeriodicSegmentalConstraint.md

```diff
@@ -0,0 +1,105 @@
+using surface rather than volume integrals.
+
+The strong form is posed over domain $\Omega$ with opposing boundary pairs $\Gamma^+$
+and $\Gamma^-$ is written in [PeriodicSegmentalConstraint.md] for the mortar method.
+The corresponding weak form is (using inner-product notation):
+
+!equation id=weak-form
+\begin{aligned}
+  (\nabla w, D \nabla u)_\Omega + \langle \llbracket w \rrbracket,\tau \llbracket u \rrbracket \rangle _{\Gamma^+} - \langle \llbracket w \rrbracket,\tau \vec{\epsilon} \cdot \llbracket \vec{x} \rrbracket \rangle _{\Gamma^+} = 0\\
+  -\langle \vec{\kappa} \cdot \llbracket \vec{x} \rrbracket, \tau \llbracket u \rrbracket \rangle _{\Gamma^+} +\langle \vec{\kappa} \cdot \llbracket \vec{x} \rrbracket, \tau \vec{\epsilon} \cdot \llbracket \vec{x} \rrbracket \rangle _{\Gamma^+} = \langle \vec{\kappa} \cdot \llbracket \vec{x} \rrbracket, \vec{\sigma} \cdot \hat{n} \rangle _{\Gamma^+}
+\end{aligned}
+
+where $\epsilon$ is the average diffusive gradient to be solved for,
+$\sigma$ is the imposed average diffusive flux, and $\lambda$ is the
+Lagrange multiplier that imposes this average constraint.
+The jump operator is defined for a single valued or vector valued field
+as $\llbracket u \rrbracket = u^{+} - u^{-}$ and
+$\llbracket \vec{x} \rrbracket = \vec{x}^{+} - \vec{x}^{-}$, respectively.
+Finally, $\tau$ is a penalty parameter to impose the constraint.
+
+The periodic constraint $\llbracket u \rrbracket = \vec{\epsilon} \cdot \llbracket \vec{x} \rrbracket$
+will only be satisfied approximately by this method, depending on the size of the penalty parameter.
+Alternatively, the periodic condition can be imposed by the Lagrange multiplier method using [PeriodicSegmentalConstraint.md] or one of the other periodic approaches in `MOOSE`.
+
+## Input File Parameters
+
+The terms in the weak form [weak-form] are handled by several different classes.
+The volume integrals are handled by [`Diffusion`](source/kernels/Diffusion.md) or
+[`MatDiffusion`](source/kernels/MatDiffusion.md). The surface term
+$\langle \llbracket w \rrbracket,\tau \llbracket u \rrbracket \rangle _{\Gamma^+}$ is computed by
+[PenaltyEqualValueConstraint.md]. The remaining four terms are handled by this class.
+
+Two of these objects are shown in the input file below:
+
+!listing test/tests/mortar/periodic_segmental_constraint/penalty_periodic_simple2d.i block=Constraints
+
+The applied macroscale diffusive flux $\sigma$ is applied as the `sigma` vector via an auxillary
+scalar. The computed macroscale diffusive gradient $\epsilon$ is assigned in a scalar variable `epsilon`.
+Both of these scalars should have the same number of components as the spatial dimension of $\Omega$.
+The volume integral of the gradient of the primary field will be constrained to $\epsilon$
+in a weak sense, depending on the size of the penalty parameter `penalty_value`.
+
+Also, the `coupled_scalar` must be assigned the same scalar as `epsilon`.
```

Comment:
The issue here is:

1. I want to let the developer change the name of the scalar variable input parameter to something "relevant" for the physics, while inside the code it will always be `kappa` to make the assembly a standard process. Also, the scalar variable should be OPTIONAL, since for some classes that want to derive off of this, we actually want to not use the scalar (important for the tensor mechanics examples I made).
2. I got the idea of how to rename parameters from one of the boundary condition files; I noticed parameters can be renamed (`addParam`) but not `addCoupledVar`.
3. Without providing `addCoupledVar`, I haven't figured out how to tell MOOSE (i.e. how to update the dependency/coupling table) to that the scalar variable is attached to the Kernel/Constraint. So, it effectively gets skipped in the Jacobian loops for off diagonal terms, and it won't appear in `const auto & coupled_scalar_vars = getCoupledMooseScalarVars();`. This seemed to happen regardless whether I used `_kappa_var_ptr(&_fe_problem.getScalarVariable())` or `_kappa_var(coupledScalar("scalar_variable"))` as the interface to the variable.


So if you know how to add the dependency of a scalar variable to an object (inside the constructor?) then that hopefully would address this.

Reply:
@lindsayad, have you made progress on the dependency issue? Sorry for the several force-pushes, but I am trying to squash some commits and merges before the final incorporation.

Perhaps the dependency could be handled similar to how the Mortar classes add the primary, secondary, and lower variables as optional. But I'm not versed enough on how to do that in the constructor.

Reply:
Yea currently the container returned by `getCoupledMooseScalarVars()` is only influenced by what is in `validParams`. So the easiest way to keep that container working for you is to keep the `"coupled_scalar"` param around. So is the reason for `"scalar_variable"` parameter is so you can rename that parameter in derived classes? If that is the sole motivator, then it might be good to add the rename capability to coupled vars as it seems that would solve 1) your desire to allow physics developers to rename the parameter 2) the duplicate parameter issue

Reply:
Yes, from what you say, adding the rename capability to coupled vars would solve both issues more easily. I guess my motivation for wanting a "real name" of the scalar variable was because there are potentially a larger variety of meanings of that variable in the coupled system. Like I saw the names for strain and stress in the tensor mechanics homogenization.

But now as I look across the `DGLowerDKernel` and `Kernel` `ScalarKernel` and `Mortar` objects, they are pretty standard at using the names "variable", "secondary_variable", and "lowerd_variable", even though the physical meaning of that variable is different depending on the PDE/weak-form.

So, maybe just retaining "scalar_variable" and not allowing renaming is more consistent with existing objects, and leaving the definition/meaning of the variable to the documentation? What do you think? Rather, it seems that specialized names/parameters are used only for off-diagonal coupled variables for the particular objects that a developer implements for a specialized physical case.

Reply:
I merged your new API, updated all source files, and removed all duplicate parameters in input files, as well as the use of `NullScalarKernel`. So this has been addressed.